core/src/solution_submission.rs

```diff
@@ -174,77 +174,104 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
-        U256::from((gas_cap.as_u128() as f64 / min_gas_price_increase_factor).floor() as u128);
-
-    let mut gas_price_estimate = U256::from(DEFAULT_GAS_PRICE);
-    let mut gas_price_factor = 1;
-    let mut result;
-    // the following block emulates a do-while loop
-    while {
-        gas_price_estimate = match gas_price_estimating.estimate_gas_price().await {
-            Ok(gas_estimate) => gas_estimate.fast,
-            Err(ref err) => {
-                log::warn!(
-                    "failed to get gas price from gnosis safe gas station: {}",
-                    err
-                );
-                gas_price_estimate
-            }
-        };
-        // Never exceed the gas cap.
-        let gas_price = std::cmp::min(gas_price_estimate * gas_price_factor, gas_cap);
+    const MIN_GAS_PRICE_INCREASE_FACTOR: f64 = 1.125 * (1.0 + f64::EPSILON);
 
-        // Submit solution at estimated gas price (not exceeding the cap)
-        result = contract
+    let effective_gas_price_cap = U256::from(
+        (gas_price_cap.as_u128() as f64 / MIN_GAS_PRICE_INCREASE_FACTOR).floor() as u128,
+    );
+    assert!(effective_gas_price_cap <= gas_price_cap);
+
+    let mut gas_price_estimator =
+        InfallibleGasPriceEstimator::new(gas_price_estimating, DEFAULT_GAS_PRICE.into());
+
+    for gas_price_increase_count in 0u32.. {
+        let estimated_price = gas_price_estimator.estimate().await;
+        let gas_price = gas_price(estimated_price, gas_price_increase_count, gas_price_cap);
+        assert!(gas_price <= gas_price_cap);
+        info!(
+            "solution submission try {} with gas price {}",
+            gas_price_increase_count, gas_price
+        );
+        let is_last_iteration = gas_price >= effective_gas_price_cap;
+        let block_timeout = if is_last_iteration {
+            None
+        } else {
+            Some(BLOCK_TIMEOUT)
+        };
+        let result = contract
             .submit_solution(
                 batch_index,
                 solution.clone(),
                 claimed_objective_value,
                 gas_price,
-                if gas_price == gas_cap {
-                    None
-                } else {
-                    Some(BLOCK_TIMEOUT)
-                },
+                block_timeout,
             )
             .await;
-
-        let gas_price_can_still_increase = gas_price < effective_gas_cap;
-        // Breaking condition for our loop
-        gas_price_can_still_increase
-            && matches!(
-                result,
-                Err(MethodError {
-                    inner: ExecutionError::ConfirmTimeout,
-                    ..
-                })
-            )
-    } {
-        // Increase gas
-        gas_price_factor *= INCREASE_FACTOR;
-        info!(
-            "retrying solution submission with increased gas price factor of {}",
-            gas_price_factor,
-        );
+        if !is_confirm_timeout(&result) || is_last_iteration {
```

Comment:
Given we don't timeout in the last iteration the or condition here is not needed, right?

Reply:
That's true. Since we always set the block timeout to None in the last iteration, getting a confirm timeout means that we cannot be in the last iteration.
There is still some merit to having this in case we make the change that we do still have a block timeout even in the last iteration or in case there unexpectedly is a ConfirmTimeout anyway (ethcontract doesn't say *there will never be a confirm timeout if block timeout is set to None*) so I'd like to keep it like this with this here explanation added as a comment.
This is also the same way the previous code worked where we could make the same assumptions to remove the `gas_price_can_still_increase` check.